### PR TITLE
CICD workflow separation

### PR DESCRIPTION
**Pull request by aholstrup1**
Separate CICD workflow into two workflows:
- _PR Build_, that runs on pull requests
- _Official Build_, that runs on pushes to certain branches (e.g. `main`).

### CODE REVIEW

File: Tests/VerifyPRChanges.Action.Test.ps1
Comment (mazhelez) [inline Tests/VerifyPRChanges.Action.Test.ps1:9]: Tests for _added_ and _deleted_ are missing.

File: Actions/VerifyPRChanges/VerifyPRChanges.ps1
Comment (mazhelez) [inline Actions/VerifyPRChanges/VerifyPRChanges.ps1:30]: There aren't any yet, but what about `.psm1` files?

File: Templates/AppSource App/.github/workflows/PullRequestHandler.yaml
Comment (mazhelez) [inline Templates/AppSource App/.github/workflows/PullRequestHandler.yaml:19]: Shouldn't this job simply have `if: ${{ github.event.pull_request.base.repo.full_name != github.event.pull_request.head.repo.full_name}}` ?

File: Templates/AppSource App/.github/workflows/PullRequestHandler.yaml
Comment (mazhelez) [inline Templates/AppSource App/.github/workflows/PullRequestHandler.yaml:37]: `environments` and `environmentCount` are probably not needed.

**Review by mazhelez — COMMENTED** [review]
Great first iteration 👍 

Needs a few touches here and there.

File: Actions/VerifyPRChanges/VerifyPRChanges.ps1
Comment (freddydk) [reply Actions/VerifyPRChanges/VerifyPRChanges.ps1:30]: We will need to allow .yml or .yaml files in other directories in the repo - and also .ps1 files.
I think we should disallow these files in .github (and subfolders) and in .AL-Go folders (where we automatically will run them)

Example: Our upcoming Test Task Recorder will deliver tests in .yaml format and also the ALDoc tool will produce documentation in .yaml format.

File: Templates/Per Tenant Extension/.github/workflows/CICD.yaml
Comment (freddydk) [reply Templates/Per Tenant Extension/.github/workflows/CICD.yaml:6]: We should not change the defaults - people can do that in settings.
If people are used to these branches, things will change over an update if we remove a branch.

File: Actions/VerifyPRChanges/VerifyPRChanges.ps1
Comment (mazhelez) [reply Actions/VerifyPRChanges/VerifyPRChanges.ps1:30]: Sounds like a separate task :)

File: Templates/AppSource App/.github/workflows/CICD.yaml
Comment (mazhelez) [reply Templates/AppSource App/.github/workflows/CICD.yaml:151]: How would it work actually? Because these are new actions. They won't exist in _microsoft/AL-Go-Actions_ until the PR is merged and then deployed.

File: Actions/CalculateArtifactNames/CalculateArtifactNames.ps1
Comment (mazhelez) [reply Actions/CalculateArtifactNames/CalculateArtifactNames.ps1:11]: Another approach would be to set an environment variable `localALGoActions` to `true` whenever we ran actions locally (e.g. in tests). That way we won't have to pollute almost every action with an extra parameter.

File: Templates/AppSource App/.github/workflows/CICD.yaml
Comment (mazhelez) [reply Templates/AppSource App/.github/workflows/CICD.yaml:11]: I believe the idea is that we cannot be sure what should and what shouldn't trigger a build. 
So, by default, we do it for every pushed change.

Here's a scenario to consider: The .md files in the projects contain documentation, so on every new build+deliver an official documentation somewhere is updated.

File: Templates/AppSource App/.github/workflows/CICD.yaml
Comment (freddydk) [reply Templates/AppSource App/.github/workflows/CICD.yaml:151]: correct - when I grab this SHA in my repo and run deploy - the new actions will be deployed and everything is fine.
Everything in microsoft/AL-Go* needs to point to microsoft/AL-Go* - else it will not be picked up by deploy/collect

File: Templates/AppSource App/.github/workflows/CICD.yaml
Comment (freddydk) [reply Templates/AppSource App/.github/workflows/CICD.yaml:11]: Good point - we should probably create a setting (like CICDPushBranches) called CICDPathsIgnore which is an array, we merge into pathsignore during Update AL-Go System Files and default that to the files that were there.

File: Actions/CalculateArtifactNames/CalculateArtifactNames.ps1
Comment (freddydk) [reply Actions/CalculateArtifactNames/CalculateArtifactNames.ps1:11]: GitHub Outputs are (by default) written to the file pointed out by $ENV:GITHUB_OUTPUT.
During tests, shouldn't we just set that environment variable to a temp filename - and then read that file to check whether the intended outcome is written?
This way we wouldn't have to have any code-change in the action to run it under tests?

File: Actions/VerifyPRChanges/VerifyPRChanges.ps1
Comment (freddydk) [reply Actions/VerifyPRChanges/VerifyPRChanges.ps1:30]: Yeah OK - we can include that in the PR for those things.

File: Actions/CalculateArtifactNames/CalculateArtifactNames.ps1
Comment (aholstrup1) [reply Actions/CalculateArtifactNames/CalculateArtifactNames.ps1:11]: Good idea :) I'll update that

File: Templates/Per Tenant Extension/.github/workflows/CICD.yaml
Comment (aholstrup1) [reply Templates/Per Tenant Extension/.github/workflows/CICD.yaml:6]: This comment isn't really related to the PR so I'll close it. I tend to agree with Maria, that I would probably also expect this to only include main and release

File: Templates/AppSource App/.github/workflows/PullRequestHandler.yaml
Comment (aholstrup1) [reply Templates/AppSource App/.github/workflows/PullRequestHandler.yaml:19]: Good point :) Thank you

File: Templates/AppSource App/.github/workflows/CICD.yaml
Comment (aholstrup1) [reply Templates/AppSource App/.github/workflows/CICD.yaml:151]: As Maria said, I'm using my own AL-GO-Actions right now because these code changes don't exist in microsoft/AL-GO :) I'll make sure to revert them back before merging the PR. 
Perhaps we should consider checking this in the gates, so we won't accidently end up merging private AL-GO-Actions into the repo

File: Templates/AppSource App/.github/workflows/CICD.yaml
Comment (aholstrup1) [reply Templates/AppSource App/.github/workflows/CICD.yaml:144]: Same as https://github.com/microsoft/AL-Go/pull/382#discussion_r1095314511

File: Templates/AppSource App/.github/workflows/PullRequestHandler.yaml
Comment (aholstrup1) [reply Templates/AppSource App/.github/workflows/PullRequestHandler.yaml:19]: Use 'If not failed'

File: Templates/AppSource App/.github/workflows/CICD.yaml
Comment (aholstrup1) [reply Templates/AppSource App/.github/workflows/CICD.yaml:11]: Reverted to defaults

File: Actions/VerifyPRChanges/VerifyPRChanges.ps1
Comment (mazhelez) [inline Actions/VerifyPRChanges/VerifyPRChanges.ps1:19]: This check is already covered in the workflow itself.

File: Templates/AppSource App/.github/workflows/CICD.yaml
Comment (freddydk) [reply Templates/AppSource App/.github/workflows/CICD.yaml:151]: The end-2-end tests will deploy the 3 repos and run the tests on them - meaning that when running the end-2-end tests on your SHA - some will use the actions in your repo.
You need to run the end-2-end tests after all references is microsoft/AL-Go-xxx
I noticed that a few tests are failing with a wrong number of actions run - I guess that is because we are running fewer actions when combining PR handler with build. This probably needs to be fixed in the tests.

File: Templates/AppSource App/.github/workflows/PullRequestHandler.yaml
Comment (freddydk) [inline Templates/AppSource App/.github/workflows/PullRequestHandler.yaml:145]: Running the end-2-end tests will use aholstrup1/AL-Go-Actions and NOT what you have in the PR here.

File: Actions/CalculateArtifactNames/CalculateArtifactNames.ps1
Comment (freddydk) [inline Actions/CalculateArtifactNames/CalculateArtifactNames.ps1:14]: I thought that if your action doesn't have any outputs defined in action.yaml, I don't think outputting anything to GITHUB_OUTPUT will have any meaning.
I saw that you defined outputs in CI/CD grabbing values from this action - does that really work?

File: Actions/TelemetryHelper.psm1
Comment (freddydk) [inline Actions/TelemetryHelper.psm1:14]: Today all the telemetry functions are used through BcContainerHelper.
We should be able to add to telemetry without containerhelper and then add DO numbers for new actions here,
Will add that as a separate refactoring workitem - we should not hold this PR up for that I think

File: Actions/ReadSettings/ReadSettings.ps1
Comment (freddydk) [inline Actions/ReadSettings/ReadSettings.ps1:147]: We should have a different way to determine this than the name of the workflow.
CI/CD is the name property in the yaml file and if we at some point in time decide to change that, we will not see this.
IMO - this value should come in a parameter

File: Actions/VerifyPRChanges/VerifyPRChanges.ps1
Comment (freddydk) [inline Actions/VerifyPRChanges/VerifyPRChanges.ps1:11]: In other actions, the token is just called token and placed first in the order.
I think we should use the same name here (unless it is a different token - and I don't think it is)

**Review by freddydk — CHANGES_REQUESTED** [review]

File: Actions/CalculateArtifactNames/CalculateArtifactNames.ps1
Comment (aholstrup1) [reply Actions/CalculateArtifactNames/CalculateArtifactNames.ps1:14]: You're right :( I discovered this yesterday as well. I can define the outputs for CalculateArtifactNames in the action.yaml but for DetermineBuildOrder this might not be possible, since our outputs names are variable e.g. projects1json, projects2json

File: Templates/AppSource App/.github/workflows/PullRequestHandler.yaml
Comment (aholstrup1) [reply Templates/AppSource App/.github/workflows/PullRequestHandler.yaml:145]: All my changes should be deployed to aholstrup1/AL-Go-Actions so it should work, no?

Comment (freddydk) [reply Templates/AppSource App/.github/workflows/PullRequestHandler.yaml:173]: Yes - it should work.
and... - I will run all end-2-end tests before release to preview anyway - but we should probably have a test, which ensures that all references to AL-Go-Actions are microsoft (as you mentioned)

File: Actions/CalculateArtifactNames/CalculateArtifactNames.ps1
Comment (freddydk) [reply Actions/CalculateArtifactNames/CalculateArtifactNames.ps1:14]: and here you need to add the outputs to the test as well (see other actions with output)

File: Actions/ReadSettings/ReadSettings.ps1
Comment (aholstrup1) [reply Actions/ReadSettings/ReadSettings.ps1]: We decided to build all projects unless the workflow is triggered by pull_request or pull_request_target

File: Actions/CalculateArtifactNames/CalculateArtifactNames.ps1
Comment (aholstrup1) [reply Actions/CalculateArtifactNames/CalculateArtifactNames.ps1:14]: Outputs added to CalculateArtifactNames

File: Templates/Per Tenant Extension/.github/workflows/PullRequestHandler.yaml
Comment (freddydk) [inline Templates/Per Tenant Extension/.github/workflows/PullRequestHandler.yaml:60]: use DO0104 and add that to telemetryhelper as PullRequestHandler

File: Templates/Per Tenant Extension/.github/workflows/PullRequestHandler.yaml
Comment (freddydk) [inline Templates/Per Tenant Extension/.github/workflows/PullRequestHandler.yaml:245]: use DO0104

File: Scenarios/EnablingTelemetry.md
Comment (freddydk) [inline Scenarios/EnablingTelemetry.md:37]: Add DO0104 to telemetryHelper and here (as suggested otherwere) - and add the other missing entries as well:-)

**Review by freddydk — CHANGES_REQUESTED** [review]

**Comment by freddydk** [comment]
I have requested a few more changes, I am OK with merging after this.
After merging - I will still run a number of tests before deploying to preview - and then next week we can deploy to preview,

File: Actions/CalculateArtifactNames/CalculateArtifactNames.ps1
Comment (mazhelez) [inline Actions/CalculateArtifactNames/CalculateArtifactNames.ps1:6]:
```suggestion
    [Parameter(HelpMessage = "Build mode used when building the artifacts", Mandatory = $true)]
```

File: Actions/ReadSettings/ReadSettings.ps1
Comment (mazhelez) [inline Actions/ReadSettings/ReadSettings.ps1:41]: No need for the `elseif` if the `if` returns.

File: Actions/ReadSettings/ReadSettings.ps1
Comment (mazhelez) [inline Actions/ReadSettings/ReadSettings.ps1:48]: Building all projects, because...? 😕

File: Actions/ReadSettings/ReadSettings.ps1
Comment (mazhelez) [inline Actions/ReadSettings/ReadSettings.ps1:52]:
```suggestion
            Write-Host "Changes to repo settings, building all projects"
```

File: Actions/ReadSettings/ReadSettings.ps1
Comment (mazhelez) [inline Actions/ReadSettings/ReadSettings.ps1:66]:
```suggestion
                if (Test-Path -Path (Join-Path $baseFolder "$checkProject/.AL-Go/settings.json")) {
```

File: Actions/ReadSettings/ReadSettings.ps1
Comment (mazhelez) [inline Actions/ReadSettings/ReadSettings.ps1:210]: This is output already in `Get-ProjectsToBuild` so I suggest to remove one of the logging (probably the one in `Get-ProjectsToBuild`).

File: Actions/ReadSettings/ReadSettings.ps1
Comment (mazhelez) [inline Actions/ReadSettings/ReadSettings.ps1:211]: Also, is this line of logging needed?

File: Actions/VerifyPRChanges/VerifyPRChanges.ps1
Comment (mazhelez) [inline Actions/VerifyPRChanges/VerifyPRChanges.ps1:11]: So, this parameter is here because it's needed in the in PS and there `github.api_url` is not in the context, right?

I don't particularly like the approach of having parameters to sake of matching YAML to PS. Also, with a bit more complex actions (that's not just calling a PS script), this whole idea collapses. Up to you to decide what's better, but in my opinion, it's not worth polluting the actions so that we can test that they call PS with the same parameters.

File: Actions/VerifyPRChanges/VerifyPRChanges.ps1
Comment (mazhelez) [inline Actions/VerifyPRChanges/VerifyPRChanges.ps1:29]: What about everything in .github folder? Didn't we agree that's a no-go for PRs?

**Review by mazhelez — APPROVED** [review]
Left some non-blocking comments. 

After fixing the telemetry, let's merge this cutie 🚀

File: Actions/ReadSettings/ReadSettings.ps1
Comment (aholstrup1) [reply Actions/ReadSettings/ReadSettings.ps1]: Now that we only do incremental builds for PRs, this shouldn't be needed

File: Actions/VerifyPRChanges/VerifyPRChanges.ps1
Comment (aholstrup1) [reply Actions/VerifyPRChanges/VerifyPRChanges.ps1:11]: I suppose in this case it's such a simple parameter, that we could just put it in the script instead of passing it as a parameter

I agree on your point about the Yaml tests though

**Pull request merged**
